### PR TITLE
feat - JSON.ARRINDEX

### PR DESCRIPTION
**Pull request by Rits1272**
This PR adds support for `JSON.ARRINDEX`

https://redis.io/docs/latest/commands/json.arrindex/#:~:text=ARRINDEX%20returns%20an%20array%20of,see%20Redis%20serialization%20protocol%20specification.

closes: https://github.com/DiceDB/dice/issues/486

### CODE REVIEW

**Comment by Rits1272** [comment]
@JyotinderSingh have added support for `JSON.ARRINDEX`. Please have a look. TIA!

File: integration_tests/commands/async/json_arrindex_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestJSONARRINDEX(t *testing.T) {
+	conn := getLocalConnection()
+	defer conn.Close()
+
```
Comment (JyotinderSingh) [inline integration_tests/commands/async/json_arrindex_test.go:11]: Add this cleanup code here to ensure that this test's results are not influenced by other tests, and that it cleans up all changes it makes once it is done executing.
```suggestion

FireCommand(conn, "DEL key")
defer FireCommand(conn, "DEL key")

```

File: integration_tests/commands/async/json_arrindex_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+			FireCommand(conn, "DEL nonExistentKey")
+			FireCommand(conn, "DEL key")
+
+			for i, cmd := range tc.commands {
+				result := FireCommand(conn, cmd)
+				expected := tc.expected[i]
+
+				if !reflect.DeepEqual(result, expected) {
+					t.Fatalf("Assertion failed: the expected value is different than result: %s", tc.name)
+				}
```
Comment (JyotinderSingh) [inline integration_tests/commands/async/json_arrindex_test.go:93]: Make use of the assert library's DeepEqual API to perform this check.

File: internal/eval/eval.go [hunk 3]
```diff
@@ -4191,6 +4191,107 @@ func evalHRANDFIELD(args []string, store *dstore.Store) []byte {
...
+
+	if strings.Contains(args[2], `"`) {
+		// user has provided string argument
+		value = args[2]
+	} else {
+		// parse it to float since default arg type would be string
+		value, err = strconv.ParseFloat(args[2], 64)
+
+		if err != nil {
+			return diceerrors.NewErrWithMessage("invalid value provided")
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:4215]: The error message here should match what redis returns.

File: internal/eval/eval.go [hunk 4]
```diff
@@ -4191,6 +4191,107 @@ func evalHRANDFIELD(args []string, store *dstore.Store) []byte {
 	return selectRandomFields(hashMap, count, withValues)
 }
 
+func evalJSONARRINDEX(args []string, store *dstore.Store) []byte {
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:4194]: Few things to improve here:
1. we should standardise the handling of the negative indices logic here and extract it into a separate function that can be used by other methods too (please find another eval method which implements the duplicated logic and replace it with a call to the function i am writing below this.
2. Please return error messages which are exactly the same as what redis would return. The messages being returned here are inaccurate.
3. Use reflect.DeepEqual for json object comparisons. Using `==` would firstly not always give the correct response, and in the worst case can also lead to runtime panic.
4. Please add test cases (ideally as unit tests) to validate all the error paths (which cover all the three points above).

I am presenting a cleaner way to structure this code, you will however have to make a few changes to it to ensure that the error messages are consistent with redis:
```go
func evalJSONARRINDEX(args []string, store *dstore.Store) []byte {
    if len(args) < 3 || len(args) > 5 {
        return diceerrors.NewErrArity("JSON.ARRINDEX")
    }

    key := args[0]
    path := args[1]
    valueArg := args[2]

    var start, stop int
    var err error

    // Default start and stop indices
    start = 0
    stop = -1 // -1 indicates end of array

    // Parse start and stop if provided
    if len(args) >= 4 {
        start, err = strconv.Atoi(args[3])
        if err != nil {
            return diceerrors.NewErrWithMessage("invalid start index")
        }
    }

    if len(args) == 5 {
        stop, err = strconv.Atoi(args[4])
        if err != nil {
            return diceerrors.NewErrWithMessage("invalid stop index")
        }
    }

    // Retrieve the object from the database
    obj := store.Get(key)
    if obj == nil {
        return clientio.RespNIL
    }

    // Check if the object is of type JSON
    errWithMessage := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeJSON, object.ObjEncodingJSON)
    if errWithMessage != nil {
        return errWithMessage
    }

    jsonData := obj.Value

    // Check if the path specified is valid
    expr, err := jp.ParseString(path)
    if err != nil {
        return diceerrors.NewErrWithMessage("invalid JSONPath")
    }

    // Parse the value argument as JSON
    var value interface{}
    err = sonic.Unmarshal([]byte(valueArg), &value)
    if err != nil {
        // If not valid JSON, treat it as string
        value = valueArg
    }

    results := expr.Get(jsonData)
    arrIndexList := make([]interface{}, 0, len(results))

    for _, result := range results {
        if utils.GetJSONFieldType(result) == utils.ArrayType {
            arr := result.([]interface{})
            length := len(arr)

            adjustedStart, adjustedStop := adjustIndices(start, stop, length)
            if adjustedStart == -1 {
                arrIndexList = append(arrIndexList, -1)
                continue
            }

            elementFound := false
            for i := adjustedStart; i <= adjustedStop; i++ {
                if reflect.DeepEqual(arr[i], value) {
                    arrIndexList = append(arrIndexList, i)
                    elementFound = true
                    break
                }
            }
            if !elementFound {
                arrIndexList = append(arrIndexList, -1)
            }
        } else {
            arrIndexList = append(arrIndexList, nil)
        }
    }

    return clientio.Encode(arrIndexList, false)
}

// adjustIndices adjusts the start and stop indices for array traversal.
// It handles negative indices and ensures they are within the array bounds.
func adjustIndices(start, stop, length int) (int, int) {
    if length == 0 {
        return -1, -1
    }

    if start < 0 {
        start += length
    }
    if stop < 0 || stop == -1 {
        stop += length
    }

    if start < 0 {
        start = 0
    }
    if stop < 0 {
        stop = 0
    }

    if start >= length {
        return -1, -1
    }
    if stop >= length {
        stop = length - 1
    }

    if start > stop {
        return -1, -1
    }

    return start, stop
}

```

File: integration_tests/commands/async/json_arrindex_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		{
+			name:     "should return -1 with start (out of boud) and stop (out of bound) optional param provided",
+			commands: []string{"json.set key $ " + normalArray, "json.arrindex key $ 3 6 10"},
+			expected: []interface{}{"OK", []interface{}{int64(-1)}},
+		},
+		{
+			name:     "should return list of array indexes for nested json",
+			commands: []string{"json.set key $ " + nestedArray, "json.arrindex key $.arrays.*.arr 3"},
+			expected: []interface{}{"OK", []interface{}{int64(2), int64(1), int64(-1)}},
+		},
```
Comment (JyotinderSingh) [inline integration_tests/commands/async/json_arrindex_test.go:79]: Add more test cases to validate edge cases for nested json.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Requesting changes.

**Comment by JyotinderSingh** [comment]
Hi @Rits1272, could you please address the reviews? That way we can merge this PR soon.

**Comment by lucifercr07** [comment]
@Rits1272 please resolve conflicts.

**Comment by lucifercr07** [comment]
@Rits1272 please address the review comments and resolve the conflicts.

**Comment by JyotinderSingh** [comment]
@Rits1272 Are you still working on this feature?
